File: LeetCode Problems/Other/Medium/114. Flatten Binary Tree to Linked List/solution.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode *prev;
    void preorder(TreeNode* current){
        if(!current) return;
        
        if(prev){
            prev->right = current;
            prev->left = NULL;
        }

        TreeNode* right = current->right;
        prev = current;
        preorder(current->left);
        preorder(right);
    }

    void flatten(TreeNode* root) {
        prev = NULL;
        preorder(root);
    }
};

File: LeetCode Problems/Other/Medium/114. Flatten Binary Tree to Linked List/solution.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self) -> None:
        self.prev = None

    def preorder(self, current) -> None:
        if not current:
            return
        
        if self.prev:
            self.prev.left = None
            self.prev.right = current

        right = current.right
        self.prev = current
        self.preorder(current.left)
        self.preorder(right)
        
    def flatten(self, root: Optional[TreeNode]) -> None:
        self.preorder(root)